src/hal/serial.c

```diff
@@ -68,21 +68,114 @@ void serial_send(char *data, int size)
 	usart_enable_tx_dma(USART1);
 }
 
+
+/**
+ * @brief Receive data from serial.
+ *
+ * DMA2 is configured to read from USART1 (Bluetooth) a number `size` of bytes.
+ * It writes all those bytes to `data`.
+ *
+ * An interruption is generated when the transfer is complete.
+ */
+static void serial_receive(void)
+{
+  dma_stream_reset(DMA2, DMA_STREAM5);
+  dma_enable_memory_increment_mode(DMA2, DMA_STREAM5);
+  dma_set_peripheral_size(DMA2, DMA_STREAM5, DMA_SxCR_PSIZE_8BIT);
+  dma_set_memory_size(DMA2, DMA_STREAM5, DMA_SxCR_MSIZE_8BIT);
+  dma_set_priority(DMA2, DMA_STREAM5, DMA_SxCR_PL_VERY_HIGH);
+  dma_set_transfer_mode(DMA2, DMA_STREAM5, DMA_SxCR_DIR_PERIPHERAL_TO_MEM);
+  
+  dma_set_peripheral_address(DMA2, DMA_STREAM5, (uint32_t)&USART3_DR);
+  
+  dma_set_peripheral_address(DMA2, DMA_STREAM5, (uint32_t)&USART1_DR);
+  dma_set_memory_address(DMA2, DMA_STREAM5, (uint32_t)receive_buffer);
+  dma_set_number_of_data(DMA2, DMA_STREAM5, RECEIVE_BUFFER_SIZE);
+
+  dma_enable_transfer_complete_interrupt(DMA2, DMA_STREAM5);
+  dma_channel_select(DMA2, DMA_STREAM5, DMA_SxCR_CHSEL_4);
+  dma_enable_stream(DMA2, DMA_STREAM5);
+  usart_enable_rx_dma(USART1);
+  
+}
+
+
+/**
+ * @brief DMA 2 stream 7 interruption routine.
+ *
+ * Executed on serial transfer complete. Clears the interruption flag, and
+ * disables serial transfer DMA until next call to `serial_send()`.
+ *
+ * It will also release the serial transfer lock.
+ */
+void dma2_stream7_isr(void)
+{
+	if (dma_get_interrupt_flag(DMA2, DMA_STREAM7, DMA_TCIF))
+		dma_clear_interrupt_flags(DMA2, DMA_STREAM7, DMA_TCIF);
+
+	dma_disable_transfer_complete_interrupt(DMA2, DMA_STREAM7);
+	usart_disable_tx_dma(USART1);
+	dma_disable_stream(DMA2, DMA_STREAM7);
+	mutex_unlock(&_send_lock);
+}
+
+/**
+ * @brief DMA 2 stream 5 interruption routine.
+ *
+ * Executed on serial receive complete. Clears the interruption flag, and
+ * resets the receive DMA by calling `serial_receive()`.
+ *
+ * This function should never be reached. Commands should be processed before
+ * on USART idle line interruption.
+ * see streams DMA mapping in section 10.3.3 (channel selection) of stm32f4 reference manual
+ **/
+void dma2_stream5_isr(void)
+{
+	if (dma_get_interrupt_flag(DMA2, DMA_STREAM5, DMA_TCIF))
+		dma_clear_interrupt_flags(DMA2, DMA_STREAM5, DMA_TCIF);
+
+	dma_disable_transfer_complete_interrupt(DMA2, DMA_STREAM5);
+	usart_disable_rx_dma(USART1);
+	dma_disable_stream(DMA2, DMA_STREAM5);
+	// LOG_ERROR("Receive buffer is full! Resetting..."); //FIXME: check after making logging functinality
```

Comment:
LOG_ERROR is a function from an upper layer. It must not be used here, as creates an unwanted dependence.